src/sources/http_scrape/scrape.rs

```diff
@@ -0,0 +1,241 @@
+//! Generalized HTTP scrape source.
+//! Scrapes an endpoint at an interval, decoding the HTTP responses into events.
+
+use bytes::{Bytes, BytesMut};
+use chrono::Utc;
+use futures_util::FutureExt;
+use http::{response::Parts, Uri};
+use snafu::ResultExt;
+use std::collections::HashMap;
+use tokio_util::codec::Decoder as _;
+
+use crate::{
+    codecs::{Decoder, DecodingConfig},
+    config::{SourceConfig, SourceContext, SourceDescription},
+    http::Auth,
+    serde::default_decoding,
+    serde::default_framing_message_based,
+    sources,
+    tls::{TlsConfig, TlsSettings},
+    Result,
+};
+use codecs::{
+    decoding::{DeserializerConfig, FramingConfig},
+    StreamDecodingError,
+};
+use vector_config::configurable_component;
+use vector_core::{
+    config::{log_schema, LogNamespace, Output},
+    event::Event,
+};
+
+/// The name of this source
+pub(crate) const NAME: &str = "http_scrape";
+
+/// Configuration for the `http_scrape` source.
+#[configurable_component(source)]
+#[derive(Clone, Debug)]
+pub struct HttpScrapeConfig {
+    /// Endpoint to scrape events from.
+    endpoint: String,
+
+    /// The interval between scrapes, in seconds.
+    #[serde(default = "super::default_scrape_interval_secs")]
+    scrape_interval_secs: u64,
+
+    /// Custom parameters for the scrape request query string.
+    ///
+    /// One or more values for the same parameter key can be provided. The parameters provided in this option are
+    /// appended to any parameters manually provided in the `endpoint` option.
+    query: Option<HashMap<String, Vec<String>>>,
+
+    /// Decoder to use on the HTTP responses.
+    #[configurable(derived)]
+    #[serde(default = "default_decoding")]
+    decoding: DeserializerConfig,
+
+    /// Framing to use in the decoding.
+    #[configurable(derived)]
+    #[serde(default = "default_framing_message_based")]
+    framing: FramingConfig,
+
+    /// Headers to apply to the HTTP requests.
+    #[serde(default)]
+    headers: Option<HashMap<String, String>>,
+
+    /// TLS configuration.
+    #[configurable(derived)]
+    tls: Option<TlsConfig>,
+
+    /// HTTP Authentication.
+    #[configurable(derived)]
+    auth: Option<Auth>,
+}
+
+impl Default for HttpScrapeConfig {
+    fn default() -> Self {
+        Self {
+            endpoint: "http://localhost:9898/logs".to_string(),
+            query: None,
+            scrape_interval_secs: super::default_scrape_interval_secs(),
+            decoding: default_decoding(),
+            framing: default_framing_message_based(),
+            headers: None,
+            tls: None,
+            auth: None,
+        }
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
```

Comment:
Could probably split the args into a separate `Struct` but I'm not totally sure what makes sense to move, maybe just `endpoint` and `query`? 💭

Reply:
Yeah I debated having this `new()` function at all- it is actually only used for the unit/integration tests. Otherwise I would have had to make the member variables of `HttpScrapeConfig` be public, which felt wrong.

Reply:
I don't have a _good_ (IMHO) suggestion either, and we have allowed that in other places. I'm OK giving the PR a +1 if we can't think of a good/natural way to split this up.

Reply:
I just looked at what the `datadog_agent` source does, (since it also separates tests and int tests into separate files) , and they too have a `new()` function for the source config, so I think keeping that is a good idea as opposed to making the member vars public. 
So its just a question of whether to split out the params into separate structs to avoid the clipping flag , or to keep the macro I have.
I'm tending to lean towards keeping the macro, as it just seems like extra boiler plate code 🤷  , but if folks are against that macro, I can certainly extract `headers, tls, auth` into something like `HttpSettings`

Reply:
No strong opinion from me end, happy to defer to your opinion/other reviewers.

Reply:
I don't see this function having any value because it's not setting any fields nor altering any values, and you have to turn off a warning to make it work. It's also not a struct for which we want to hide the values (unlike, say, `HttpScrapeContext`) to avoid external code from altering state it shouldn't.

Reply:
The only value it provides is not having to make the member variables of the `HttpScrapeConfig` to be public.
I was mimicking the `datadog_agent` source as they too had separated integration and unit test files, so they opted to create a new() function as opposed to making the members public. Unless there is an alternative option ...
What are your thoughts on that aspect of it ?

Reply:
Avoiding public members in structs where we don't want internal state touched is a good idea. The config structs are somewhat of a special case, as the members aren't really private state (I mean, users set the fields through the configuration file) so I see no reason for them not to be public.